tools/packaging/packaging.strings.psd1

```diff
@@ -11,6 +11,14 @@ if [ ! -f /etc/shells ] ; then
 else
     grep -q "^{0}$" /etc/shells || echo "{0}" >> /etc/shells
 fi
+if [ -f /lib64/libssl.so.1.1 ] ; then
```

Comment:
See https://github.com/PowerShell/PowerShell/issues/15029#issuecomment-800690344 you can't just link the `libssl` and `libcrypto` symlinks to the OpenSSL 1.1.x version and just expect it to work. It does allow the library to load but causes a hard fault in the process as soon as `libmi` tries to access any of the OpenSSL libraries in that. OpenSSL 1.1.x and 1.0.x (what `libmi` that is shipped by PowerShell is linked to) are not API and ABI compatible.

You need to either link it to OpenSSL 1.0.x and document the `compat-openssl10` (other whatever package provides OpenSSL 1.0.x on the distro) requirement for WSMan work or ship a recompiled `libmi` against OpenSSL 1.1.x so it can use a modern OpenSSL version.

Reply:
> ship a recompiled libmi against OpenSSL 1.1.x

If I remember correctly this is blocked by external team.

------------------

We could try to load libmi and the old openssl to alternative ALC. I hope AssemblyLoadContext.LoadUnmanagedDll() will do this. If yes it is not too huge work. /cc @rjmholt

Reply:
There’s a long history behind it, PowerShell can always just recompile the libraries from the same source and ship it. It’s not the team maintains it anyway as the source is missing a lot of security features and contains bugs that just aren’t focused on. But this has been commented on quite a lot of times and ultimately the team isn’t  interested in doing this. That doesn’t mean the package should point the link to the wrong OpenSSL version.

> I hope AssemblyLoadContext.LoadUnmanagedDll() will do this

I don’t think so, the linking happened at the unmanaged layer. PowerShell isn’t loading libcrypto, libssl, or even libmi directly. It’s loading libpsrpclient and whatever the native loading mechanism is loading libmi/libssl.

Ultimately you just can’t try and use libssl from 1.1.x when the calling code is expecting the symbols and API from the 1.0.x version. It might be enough to get libmi to load but as soon as you try and open a https connection it will crash the entire process. If the goal is to try and lessen the dependency on OpenSSL 1.0.x for WSMan then this is the wrong thing to do. It’s still a dependency and now you are introducing process crashes when it tried to access libssl.

Reply:
> I don’t think so, the linking happened at the unmanaged layer. PowerShell isn’t loading libcrypto, libssl, or even libmi directly. It’s loading libpsrpclient and whatever the native loading mechanism is loading libmi/libssl.

In that case, it doesn't matter at all - it didn't work before and won't work now - nothing has changed.

Reply:
It does still matter, if we look at the PowerShell codebase we can see the various PInvoke definitions for the WSMan api are being called through [libpsrpclient](https://github.com/PowerShell/PowerShell/blob/2784add414393b0ecab798055cfbac0d363c1c1e/src/System.Management.Automation/engine/remoting/fanin/WSManNativeAPI.cs#L2339). On non-Windows this is essentially a build of https://github.com/PowerShell/psl-omi-provider and is designed to replicate the Windows WSMan layer and is dynamically linked to `libmi` which is where the actual WSMan client is implemented.

The crux of this problem is what `libmi` is dyamically linked to. In both the version provided for EL7 and EL8 we can see that they are both linked against OpenSSL 1.0.x.

```
[root@45affdb15cca /]# ldd /opt/microsoft/powershell/7/libmi.so
ldd: warning: you do not have execution permission for `/opt/microsoft/powershell/7/libmi.so'
        linux-vdso.so.1 (0x00007ffde7395000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff4980fc000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007ff497ef8000)
        libpam.so.0 => /lib64/libpam.so.0 (0x00007ff497ce8000)
        libssl.so.1.0.0 => /opt/microsoft/powershell/7/libssl.so.1.0.0 (0x00007ff497a54000)
        libcrypto.so.1.0.0 => not found
        libc.so.6 => /lib64/libc.so.6 (0x00007ff497691000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff49831c000)
        libaudit.so.1 => /lib64/libaudit.so.1 (0x00007ff497467000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007ff496f81000)
        libz.so.1 => /lib64/libz.so.1 (0x00007ff496d6a000)
        libcap-ng.so.0 => /lib64/libcap-ng.so.0 (0x00007ff496b64000)
```

_Note: `libssl` and `libcrypto` have the explict `1.0.0` extension to denote the version._

This PR (and what's currently in master) creates 2 symbolic links in the PowerShell install directory.

|Distribution|libcrypto.so.1.0.0 target|libssl.so.1.0.0 target|
|-|-|-|
|EL7|/lib64/libcrypto.so.10|/lib64/libssl.so.10|
|EL8|/lib64/libcrypto.so.1.1.1|/lib64/libcrypto.so.1.1|
|Debian 9|/usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2|/usr/lib/x86_64-linux-gnu/libssl.so.1.0.2|
|Debian 10+|/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1|/usr/lib/x86_64-linux-gnu/libssl.so.1.1|

This symbolic link means that when the `libmi` library is loaded and the loader goes through the linked dependencies and finds `libssl.so.1.0.0` and `libcrypto.so.1.0.0` it will select the symbolic link in the PowerShell directory that the points to the proper proper that is specific for that distribution.

If any of these dependencies (as denoted by `ldd`) cannot be found then `libmi` will fail to load which also means `libpsrpclient` cannot be loaded by PowerShell causing the error

> This parameter set requires WSMan, and no supported WSMan client library was found. WSMan is either not installed or unavailable for this system.

From what I can gather the intent of this PR and what the current package symlinks target is to ensure that `libssl` and `libcrypto` libraries are pointing to an actual library specific for that distribution so `libmi` can be loaded and used by `libpsrpclient`. The problem with this approach is that `libmi` has been compiled against OpenSSL 1.0.x and not 1.1.x so anything in `libmi` that calls or references a symbol in `libssl` or `libcrypto` either may not exist (API) or the differ in how it is meant to be called (ABI) causing a failure and crashing the process with

> pwsh: symbol lookup error: /opt/microsoft/powershell/7/libmi.so: undefined symbol: SSL_library_init

This will occur for anything that tries to connect over a `https` endpoint over WSMan, like Exchange Online.

Ultimately there are a few options available:

* Continue as is and create a symlink pointing to OpenSSL 1.1.x on the distributions that default to that version
  * It is more likely WSMan will work on `http` out of the box but using `https` will crash the entire process
  * Gives a false indication that the WSMan libs work with OpenSSL 1.1.x and not the old and unsupported OpenSSL 1.0.x
  * The above doesn't matter as much because anything TLS related will just cause a crash but it is still an incorrect assumption
* Have the symlinks point to the OpenSSL 1.0.x path for all distributions
  * EL7 OpenSSL package is based on OpenSSL 1.0.2 so this is the out of box experience
  * EL8 has the `compat-openssl10` package which provides the OpenSSL 1.0.2 libs
  * Debian 9 has the `libssl1.0.2` package
  * Debian 10 doesn't have the `libssl1.0.2` package but you can technically just install it from the Debian 9 repo
  * Ubuntu is a bit harder but it [seems to be possible to install it](https://askubuntu.com/questions/989086/stuck-downgrading-libssl-on-ubuntu-16-04)
  * This ensures that TLS works with WSMan and people are explicitly aware that they are using an older and unsupported OpenSSL version
* Provide a new build of `libmi` with PowerShell that is compiled against OpenSSL 1.1.x
  * While the `omi` team doesn't support PowerShell I don't see why this stops PowerSell from using their "official" releases https://github.com/microsoft/omi/releases/tag/v1.6.6-0
  * They include a "universal" build for both OpenSSL 1.0.x and 1.1.x allowing it to work on both older and newer distributions
  * The PS team seems quite reluctant to using this but honestly I see this as a win win; PowerShell now uses OpenSSL 1.1.x which supports newer cipher suites and protocols and it works out of the box for more distributions
  *  It's not like the current distributed `libmi` in PowerShell is accepting any bugfixes so why not just use what the OMI team has already released which works with newer versions
  * There is a complication with `macOS` but that's a whole other story and I don't see why it should bring down Linux

I really hope the 3rd option is chosen, I just don't see any reason why PowerShell can't just use what is already provided by the OMI team instead of continuing to require OpenSSL 1.0.x to use WSMan over HTTPS.

Reply:
@jborean93 Thanks for great post! But... I know that you known that MSFT does not support this scenario with libmi, they have officially announced that they will be removing libmi at all. (So DSC will be replaced by classes, and remoting will be transformed into a subsystem so that we can create modern alternatives.)

The PR is only for making universal Unix package and here is nothing about fixing any scenario.

Reply:
The goal here is not to improve our support for libmi.

Reply:
So you are going to create completely new code that just replicates the existing, and known to be wrong, behaviour rather than fix it at the same time. Seems like a waste of effort but I’m tired of trying to advocate this be fixed so I’m giving up.

Reply:
@jborean93 I agree with you that a working solution should have been made first 4 years ago, but this has not been done. I am at a loss and can only explain this by political wars. But this is no longer important, since the time has already passed. For now, we only have to insist on the implementation of a new remoting subsystem. This will untie the hands of both the community and MSFT itself.